sw/device/lib/base/hardened_memory.h

```diff
@@ -0,0 +1,89 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_HARDENED_MEMORY_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_HARDENED_MEMORY_H_
+
+/**
+ * @file
+ * @brief Hardened memory operations for constant power buffer manipulation.
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/hardened.h"
+#include "sw/device/lib/base/macros.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Copies 32-bit words between non-overlapping regions.
+ *
+ * Unlike `memcpy()`, this function has important differences:
+ * - It is significantly slower, since it mitigates power-analysis attacks.
+ * - It performs operations on 32-bit words, rather than bytes.
+ * - It returns void.
+ *
+ * Input pointers *MUST* be 32-bit aligned, although they do not need to
+ * actually point to memory declared as `uint32_t` per the C aliasing rules.
```

Comment:
Personally I like taking a well-typed pointer, since it forces callers to actually ensure alignment; this isn't a function to call willy-nilly. The cast itself isn't a strict aliasing violation, since we never dereference through `uint32_t&` objects inside the function body.